proposals/0110-exponential-fee-for-write-lock-accounts.md

```diff
@@ -0,0 +1,147 @@
+  `alpha = 2 / (N+1)`.
+- Pricing Algorithm:
+  - Adjusts write-lock *cost rate* based on an account's EMA *compute-unit
+  utilization*.
+  - For each block, if an account's EMA *compute-unit utilization* is more than
+  half of its max limit, its write-lock *cost rate* increases by X%. If it's
+  below half, the *cost rate* decreases by X%.
+  - For V0, Initial write-lock cost rate is `1000 micro-lamport/CU`; and cost
+  rate set to 1%.
+- Calculate *Write Lock Fee*:
+  - Fee required to write-lock an account is calculated by multiplying the
+  write-lock *cost rate* by the transaction's requested CU.
+
+### Detailed Design
+
+- Initialization and Inheritance:
+  - Bank initializes an empty account_write_lock_fee_cache, an LRU Cache of
+  {account_pubkey, compute-unit-pricer}.
+  - Child banks inherit the parent's cache.
+- Transaction Fee Calculation:
+  - Calculate write-lock fee for each account a transaction needs to write,
+  summing up to be its *write lock fee*. This, along with signature fee and
+  priority fee, constitutes the total fee for the transaction.
+  - Leader checks fee payer's balance before scheduling the transaction.
+- Cost Tracking:
+  - Cost_tracker tracks CUs for the current block and each write-locked account
+  as-is;
+  - Ensuring cost tracking is enabled at the replay stage.
+- End of Block Processing:
+  - Identify write-locked accounts with *compute-unit utilization* > half of
+  account max CU limit. Add/update bank's account_write_lock_fee_cache. 
+  - Evicting cheapest account before add new "hot" accounts into LRU cach;
+  - LRU cache has capacity set to 2* worst case eviction per block to prevent
+  cache attack.
+  - For v0, cache capacity set to 2048, as:
+    - Max number of tansactions with account 6M CU = 48M/6M = 8;
+    - Max number of accounts per tx: 128;
+    - worst case per block: 128 * 8 = 1024;
+    - 2 times worst case: 2048;
+- Fee Handling:
+  - Collected write-lock fees are 100% burnt.
+  - Collected priority fees are 100% rewarded.
+  - Collected signature fees are 50% burnt, 50% rewarded.
+
+### Other Considerations
+
+- Users may need new instruction to set a maximum write-lock fee for transaction
+- Consider tooling for wallets/simulators to query "min/max write lock fee."
+- Acknowledge read lock contention, deferring EMA fee implementation for read locks.
+- In the future, a percentage of collected write-lock-fee could be deposited
+to an account, allowing dApps to refund cranks and other service providers.
+This decision should be done via a governance vote.
+
+## Impact
+
+- Rate of successful CU inclusion in a block is expected to increase, reducing
+failed transactions.
+- Transactions writing to contentious accounts will experience increased fees,
+particularly during congestion.
+- DeFi arbitrage traders will need to adjust strategies to account for the
+heightened fees.
+
+## Security Considerations
+
+none
```

Comment:
Can potentially slow the leader by spamming it with transactions using 256 accounts in an ALT all write locked.

Reply:
Good point, this isn't unique to this proposal tho. ALT needs to be loaded anyway, the proposal adds LRU cache lookup

Reply:
max account locks is only 64. Scheduler should use locks in the equation however

Reply:
Feature activation pending right?
9LZdXeKGeBV6hRLdxS1rHbHoEUsKqesCC2ZAPTPKJAbK | inactive                | NA              | increase tx account lock limit to 128 #27241

Reply:
According to [feature schedule](https://github.com/solana-labs/solana/wiki/Feature-Gate-Activation-Schedule), it is blocked atm.